### PR TITLE
fix: fix `setindex!` not erroring appropriately for integrators

### LINKED ISSUE #653: Inconsistent behaviour when changing integrator variable value
In one case, using `oint[X]` works while `setu(oint, X)(oint, 10.0)` throws an error. Presumably, the first one should throw an error as well?

```julia
using  ModelingToolkit, OrdinaryDiffEq
using ModelingToolkit: t_nounits as t, D_nounits as D

@parameters p
@variables X(t)

eq = D(X) ~ p - X
@mtkbuild osys = ODESystem([eq], t)
oprob = ODEProblem(osys, [X => 0.1], (0.0, 1.0), [p => 1.0])

oint = init(oprob, Tsit5())
oint[X] = 1.0 # Works.
setu(oint, X)(oint, 10.0) # ERROR: Integrator state cannot be reset unless it is initialized with save_everystep=false

oint = init(oprob, Tsit5(); save_everystep=false)
oint[X] = 1.0 # Works.
setu(oint, X)(oint, 10.0) # Works.
```

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Close #653 

## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

File: test/downstream/integrator_indexing.jl [hunk 1]
```diff
@@ -367,3 +367,19 @@ setp(sys, p)(integrator, [4, 5, 6])
...
+@parameters p
+@variables X(t)
+
+eq = D(X) ~ p - X
+@mtkbuild osys = ODESystem([eq], t)
+oprob = ODEProblem(osys, [X => 0.1], (0.0, 1.0), [p => 1.0])
+
+oint = init(oprob, Tsit5())
+@test_throws ErrorException oint[X]=1.0
+@test_throws ErrorException setu(oint, X)(oint, 10.0)
```
Comment (ChrisRackauckas) [inline test/downstream/integrator_indexing.jl:381]: Why would this error?

Comment (AayushSabharwal) [reply test/downstream/integrator_indexing.jl:381]: https://github.com/SciML/OrdinaryDiffEq.jl/blob/258bbbbc7458b66c14fe4a03054252d176c9dea8/src/integrators/integrator_interface.jl#L480-L481

Comment (ChrisRackauckas) [reply test/downstream/integrator_indexing.jl:381]: Changing out the `u` vector is different from mutating it. These functions should just mutate `u`.

Comment (AayushSabharwal) [reply test/downstream/integrator_indexing.jl]: Fixed now

**Pull request merged**